src/scripts/make_grey_table.py

```diff
@@ -0,0 +1,295 @@
+        bb = fac[n] / (1.0 + np.exp(eta)) * (energy[:, None, None, None]) ** 3
+        # Integrate over energy to get the baseline interaction rates
+        bb_int = np.trapz(bb, x=energy, axis=0)
+        
+        # Calculate absorption rates
+        absorption = bb * nulib_table["absorption"][:, n, :, :, :]
+        absorption_int = np.trapz(absorption, x=energy, axis=0)
+        table[n, 0, :, :, :] = absorption_int.T
+        table[n, 1, :, :, :] = np.divide(absorption_int, bb_int, out=np.zeros_like(absorption_int), where=bb_int != 0).T
+
+        # Calculate emission rates
+        emiss = np.zeros_like(absorption)
+        for i in range(len(energy)):
+            emiss[i, :, :, :] = absorption[i, :, :, :] / energy[i]
+        emiss_int = np.trapz(emiss, x=energy, axis=0)
+        table[n, 3, :, :, :] = emiss_int.T
+
+        # Calculate scattering rates
+        scatter = bb * nulib_table["scatter"][:, n, :, :, :]
+        scatter_int = np.trapz(scatter, x=energy, axis=0)
+        table[n, 2, :, :, :] = np.divide(scatter_int, bb_int, out=np.zeros_like(scatter_int), where=bb_int != 0).T
+
+    return table
+
+
+def generate_and_save_table(nulib_file: str,eos_file: str,output_file: str,
+                            num_temperatures: int = 12,temperature_range: tuple = (0.1, 16)) -> None:
+    """
+    Generates a 4D table for grey neutrino interactions over a range of temperatures and saves it to an HDF5 file.
+    The table integrates neutrino interaction data (absorption, emission, and scattering rates) over energy, 
+    assuming neutrinos follow a Fermi-Dirac distribution. It consists of two parts:
+    
+    1. Equilibrium interactions: Assumes the fluid temperature and the neutrino temperature are the same. 
+    
+    2. Non-equilibrium interactions: Allows for a range of neutrino temperatures, distinct from the fluid temperature. 
+       This section is generated by sampling over specified neutrino temperatures, determined by num_temperatures 
+       and temperature_range, to account for scenarios where neutrino and fluid temperatures differ.
+    
+
+    Parameters:
+    - nulib_file (str): Path to the NuLib table file.
+    - eos_file (str): Path to the EOS table file.
+    - output_file (str): Path and filename for the output HDF5 file.
+    - num_temperatures (int): Number of temperature points to generate.
+    - temperature_range (tuple): The minimum and maximum temperatures (in MeV) for the logspace generation.
+
+    Returns:
+    - None: The function does not return a value but writes data to an HDF5 file.
+    """
+    # Read NuLib and EOS tables
+    nulib_table = read_nulib_table(nulib_file)
+    eos_table = read_eos_table(eos_file)
+
+    # Generate log-spaced neutrino temperature array
+    log_temperatures = np.logspace(np.log10(temperature_range[0]), np.log10(temperature_range[1]), num=num_temperatures)
+
+    # Initialize the 4D table array
+    table4d = np.zeros([3, 4, 82, 100, 56, len(log_temperatures)])
```

Comment:
This hard coded for a particular table

Reply:
Changed to account for the actual size of the original table. The two first dimensions have to be 3 and 4, the script will not work as is for more than 3 species.